### ISSUE TITLE
Rounding of product prices is different then rounding of tax prices

### ISSUE THREAD

**Issue by mikkamp** [woocommerce/woocommerce#23835]
**Describe the bug**
When we set the option to enter prices inclusive of tax. It means all the product prices exclusive of tax will be calculated depending on the tax rate for the shop base location.

To influence the way taxes are rounded we can set the option to round per line item or round at subtotal level. That works for the taxes but it doesn't apply this same logic to the product totals. The product totals will always be rounded per line item. The reason this is an issue here is because the product totals (exclusive of tax) are calculated values based on the tax (since we have prices set inclusive of tax).

So if we are influencing the way the rounding gets done for only the taxes but not for the product prices (exclusive of tax) then we get some rounding discrepancies.

**To Reproduce**
1. Set a tax rate of 19% (use this tax rate for the store location)
2. Create a product with a price of 95
3. Create a product with a price of 11
4. Set taxes to the following:
  - prices inclusive of tax
  - rounding at subtotal level
  - display prices in shop and cart including taxes
5. Add the two products to the cart and choose a location which applies the same 19% tax
6. Observe that the total cost is 105.99 which includes a tax of 16.92

**Expected behavior**
What I would have expected in the above scenario was for the total cost to be 106 which includes a tax of 16.92.

The reason it arrives at the total of 105.99 is because of the following calculations:

```
Product 1: 
95 / 1.19 = 79.8319 cost excluding taxes
95 - ( 95 / 1.19 ) = 15.16806723 tax

Product 2:
11 / 1.19 = 9.2437 cost excluding taxes
11 - ( 11 / 1.19 ) =  1.75630252 tax

Product totals (rounded at line item):
79.83 + 9.24 = 89.07

Tax totals (rounded at subtotal because setting is enabled):
15.16806723 + 1.75630252 = 16.92436975

Product totals + tax totals:
89.07 + 16.92436975 = 105.99
```

If instead we would round the product totals the same as the tax setting, we would get:

```
Product totals (rounded at line item):
79.8319 + 9.2437 = 89.0756

Tax totals (rounded at subtotal because setting is enabled):
15.16806723 + 1.75630252 = 16.92436975

Product totals + tax totals:
89.0756 + 16.92436975 = 105.99996975 rounded to 106
```

Changing the behaviour would be more consistent with how the taxes apply rounding, but it might have side effects for other parts like discounts per line item. To confirm this I adjusted this behaviour in the following two locations:

https://github.com/woocommerce/woocommerce/blob/3.6.4/includes/class-wc-cart-totals.php#L685
https://github.com/woocommerce/woocommerce/blob/3.6.4/includes/class-wc-cart-totals.php#L743

I added the condition `$this->round_at_subtotal()` and calculated the totals without applying the round function. And then ran the unit tests. This showed the failure of the following test:

```
There was 1 failure:

1) WC_Tests_Cart::test_cart_get_total_issue_21021
Failed asserting that '87.91' matches expected 87.9.

/www/wp-content/plugins/woocommerce/tests/unit-tests/cart/cart.php:181

FAILURES!
Tests: 1406, Assertions: 5237, Failures: 1.
```

Which points to the following issue: https://github.com/woocommerce/woocommerce/issues/21021

My main questions / concerns would be around the setting to round at subtotal level. If we do choose to set that then it would make sense for it to apply to the product total calculations as well. Especially since they are calculated values when we enter prices including tax.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>

```
WordPress 5.2.1
WooCommerce 3.6.4
```
</details>

**Comment by johfoll** [comment]
Hi, I'm new here but also have this issue and found that in addition to modifying the two lines mentioned above (although without adding the condition `$this->round_at_subtotal()`), I also modified 
https://github.com/woocommerce/woocommerce/blob/3.6.4/includes/class-wc-cart-totals.php#L591
to remove the rounding, so it becomes:
`$taxes[ $rate_id ] += $rate;`
and
https://github.com/woocommerce/woocommerce/blob/3.6.4/includes/class-wc-cart-totals.php#L734
also removing the rounding, so it becomes:
`$merged_subtotal_taxes[ $rate_id ] += $rate;`

I found that this seems to fix the total and tax amount for me, regardless of which rounding setting is enabled. It also seems to give the correct amounts when used with a coupon, but I don't have the unit tests so can't comment on those, unless someone would like to test this for me? :)

**Comment by mikkamp** [comment]
Running the unit tests is explained a bit more here: https://github.com/woocommerce/woocommerce/wiki/How-to-set-up-WooCommerce-development-environment#setting-up-unit-tests

I tested those two changes you suggested. It seems it just removes any additional rounding to make it happen as late as possible when the option round on subtotal is set.

The same unit test still fails though.

**Comment by johfoll** [comment]
Cool, thanks for doing that, and thanks for the link! Sorry, spoke too soon with my suggestion. To add to your example calculations, I would also suggest that the product totals should be rounded to the same number of decimal places as the tax totals. Summing amounts that are rounded to the same number of places cancel each other out, and give a precise result. For example, rounding everything to say, 4 decimal places, initially, would give the following:

```
Product totals (NOT rounded to 2 decimal places at line item, just rounded to 4 places initially to keep some precision):
79.8319 + 9.2437 = 89.0756

Tax totals (rounded at subtotal because setting is enabled. NOT rounded to 2 decimal places at line item, just rounded to 4 places initially to keep some precision):
15.1681 + 1.7563 = 16.9244

Product totals + tax totals:
89.0756 + 16.9244 = 106 exactly
```
Similarly, with 'round at subtotal' disabled, I would expect the total to be the same as the above, but the split between tax totals and product totals would be a little different.

```
Product totals ( rounded to 2 decimal places at line item):
79.83 + 9.24 = 89.07

Tax totals (rounded to 2 decimal places at line item):
15.17 + 1.76 = 16.93

Product totals + tax totals:
89.07 + 16.93 = 106 exactly
```

**Comment by reykjalin** [comment]
2108541-zen seems to be having the same issue even though they claim `Rounding at the subtotal level` is disabled.

I can reproduce the problem on their website, but since we have limited information from the customer it's hard to confirm whether this is the _exact_ same problem, or a variation of this same rounding logic complication.

**Comment by vedanshujain** [comment]
Possibly #24024 will fix this as well, because of late rounding of item values.

**Comment by claudiosanches** [comment]
@mikkamp can you please help us test it? Seems like is already fixed by #24024

**Comment by mikkamp** [comment]
I tested fix #24024 and I can confirm that it fixes the issue which was reported here.

With the fix in place the result of the totals are:
106 which includes a tax of 16.92
